File: 백준/Silver/1654. 랜선 자르기/랜선 자르기.cc
#include <iostream>
#include <algorithm>
using namespace std;
int arr[10001];
int K, N;
long long high, low, mid, ans;
int main()
{
    cin >> K >> N;
    for (int i = 0; i < K; i++)
    {
        cin >> arr[i];
    }
    low = 1;
    high = *max_element(arr, arr + K);
    ans = 0;
    while (low <= high)
    {
        mid = (low + high) / 2;
        int cnt = 0;
        for (int i = 0; i < K; i++)
        {
            cnt += arr[i] / mid;
        }
        if (cnt >= N)
        {
            low = mid + 1;
            ans = max(ans, mid);
        }
        else
        {
            high = mid - 1;
        }
    }
    cout << ans << "\n";
}

File: 백준/Silver/1654. 랜선 자르기/랜선 자르기.py
import sys
input = sys.stdin.readline
n, k = map(int, input().split())
arr = [int(input()) for _ in range(n)]


left = 1
right = max(arr)
ans = 0

while(left <= right):
    mid = (left+right)//2;
    cnt = 0
    for i in arr:
        cnt = cnt + (i//mid)
    if(cnt >= k):
        left = mid+1
        ans = max(ans,mid)
    else:
        right = mid -1

print(ans)